src/VisualStudio/Core/Def/Implementation/CodeLens/CodeLensCallbackListener.cs

```diff
@@ -0,0 +1,134 @@
+    /// </summary>
+    [Export(typeof(ICodeLensCallbackListener))]
+    internal class CodeLensCallbackListener : ICodeLensCallbackListener
+    {
+        private const int DefaultMaxSearchResultsValue = 99;
+
+        private const string CodeLensUserSettingsConfigPath = @"Text Editor\Global Options";
+        private const string CodeLensMaxSearchResults = nameof(CodeLensMaxSearchResults);
+
+        private readonly VisualStudioWorkspaceImpl _workspace;
+        private readonly IServiceProvider _serviceProvider;
+        private readonly JoinableTaskContext _joinableTaskContext;
+
+        private int _maxSearchResults = int.MinValue;
+
+        [ImportingConstructor]
+        public CodeLensCallbackListener(
+            SVsServiceProvider serviceProvider,
+            JoinableTaskContext joinableTaskContext,
+            VisualStudioWorkspaceImpl workspace)
+        {
+            _serviceProvider = serviceProvider;
+            _joinableTaskContext = joinableTaskContext;
+            _workspace = workspace;
+        }
+
+        public async System.Threading.Tasks.Task SynchronizePrimaryWorkspaceAsync(CancellationToken cancellationToken)
+        {
+            var client = await _workspace.TryGetRemoteHostClientAsync(cancellationToken).ConfigureAwait(false);
+            if (client == null)
+            {
+                throw new InvalidOperationException("remote host doesn't exist");
+            }
+
+            await _workspace.SynchronizePrimaryWorkspaceAsync(_workspace.CurrentSolution, cancellationToken).ConfigureAwait(false);
+        }
+
+        public async Task<string> GetHostGroupIdAsync(CancellationToken cancellationToken)
+        {
+            // in VS host, RemoteHostClient is always ServiceHubRemoteHostClient
+            var client = (ServiceHubRemoteHostClient)await _workspace.TryGetRemoteHostClientAsync(cancellationToken).ConfigureAwait(false);
+            if (client == null)
+            {
+                throw new InvalidOperationException("remote host doesn't exist");
+            }
+
+            return client.HostGroupId;
+        }
+
+        public async Task<int> GetMaxResultCapAsync(CancellationToken cancellationToken)
+        {
+            await EnsureMaxResultAsync(cancellationToken).ConfigureAwait(false);
+
+            return _maxSearchResults;
+        }
+
+        public Guid GetProjectId(Guid projectGuid, CancellationToken cancellationToken)
+        {
+            if (TryGetProjectId(_workspace, projectGuid, out var projectId))
+            {
+                return projectId.Id;
+            }
+
+            return Guid.Empty;
+        }
+
+        private async System.Threading.Tasks.Task EnsureMaxResultAsync(CancellationToken cancellationToken)
+        {
+            if (_maxSearchResults != int.MinValue)
+            {
+                return;
+            }
+
+            await _joinableTaskContext.Factory.SwitchToMainThreadAsync(cancellationToken);
```

Comment:
ick... is this really necessary?  can we not use a free-threaded options system here?

Reply:
Vs doesn't have one.